sigstore/_internal/sct.py

```diff
@@ -16,75 +16,124 @@
+def _pack_digitally_signed(
+    sct: SignedCertificateTimestamp,
+    cert: Certificate,
+    issuer_key_hash: bytes,
+) -> bytes:
     """
     The format of the digitally signed data is described in IETF's RFC 6962.
-
-    1 SCT Version
-    1 Signature Type
-    8 Timestamp
-    2 Entry Type
-    3 Certificate Length
-    X Certificate Data
-    2 Extensions Length
     """
 
     # The digitally signed format requires the certificate in DER format.
-    cert_der: bytes = cert.public_bytes(encoding=serialization.Encoding.DER)
+    cert_der: bytes = cert.tbs_precertificate_bytes
 
-    # The length should then be split into three bytes.
+    # The length is a u24, which isn't directly supported. So we have to
+    # decompose it into 3 bytes.
     unused, len1, len2, len3 = struct.unpack(
         "!4B",
         struct.pack("!I", len(cert_der)),
     )
     if unused:
         raise InvalidSctError(f"Unexpectedly large certificate length: {len(cert_der)}")
 
+    # No extensions are currently specified.
+    if len(sct.extension_bytes) != 0:
+        raise InvalidSctError("Unexpected trailing extension bytes")
+
+    # NOTE(ww): This is incorrect for non-embedded SCTs, since it unconditionally
+    # embeds the issuer_key_hash.
+
     # Assemble a format string with the certificate length baked in and then pack the digitally
     # signed data
-    pattern = "!BBQhBBB%ssh" % len(cert_der)
+    # fmt: off
+    pattern = "!BBQH32sBBB%ssH" % len(cert_der)
     data = struct.pack(
         pattern,
-        sct.version.value,
-        0,  # Signature Type
-        int(sct.timestamp.timestamp() * 1000),
-        0,  # Entry Type
-        len1,
-        len2,
-        len3,
-        cert_der,
-        len(sct.extensions),
+        sct.version.value,                      # sct_version
+        0,                                      # signature_type (certificate_timestamp(0))
+        int(sct.timestamp.timestamp() * 1000),  # timestamp (milliseconds)
+        sct.entry_type.value,                   # entry_type (x509_entry(0) | precert_entry(1))
+        issuer_key_hash,                        # issuer_key_hash[32]
+        len1,                                   # \
+        len2,                                   # | opaque TBSCertificate<1..2^24-1> OR
+        len3,                                   # | opaque ASN.1Cert<1..2^24-1>
+        cert_der,                               # /
+        len(sct.extension_bytes),               # extensions (opaque CtExtensions<0..2^16-1>)
     )
+    # fmt: on
 
     return data
 
 
+CERTIFICATE_TRANSPARENCY_OID = ObjectIdentifier("1.3.6.1.4.1.11129.2.4.4")
+
+
+def _is_preissuer(issuer: Certificate) -> bool:
+    ext_key_usage = issuer.extensions.get_extension_for_class(ExtendedKeyUsage)
+    return CERTIFICATE_TRANSPARENCY_OID in ext_key_usage.value
+
+
+def _get_issuer_cert(chain: List[Certificate]) -> Certificate:
+    issuer = chain[0]
```

Comment:
chain starts with the intermediate/issuing CA cert? Just want to check that chain doesn't start with the leaf

Reply:
I believe so, but will confirm. Interestingly things *are* working for @di, just not locally for me.

Reply:
Which environment are you testing against, staging? And using the correct verification key?

Reply:
> Which environment are you testing against, staging? And using the correct verification key?

Yep, staging, and we confirmed that we both have the correct key. @di is able to use both the OAuth and ambient flows correctly, so it's possible there's just something pathologically wrong with my local environment and/or a very subtle bug in OpenSSL that's only manifesting for me.